File: framework/languages/java/samples/java/ShoppingMall/Server/CommerceApi/src/main/java/systems/zlink/samples/shoppingmall/server/commerceapi/handlers/CreatePendingMappingHandler.java
package systems.zlink.samples.shoppingmall.server.commerceapi.handlers;

import systems.zlink.framework.channels.ZLinkRequestContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.shoppingmall.server.configuration.CommerceStore;
import systems.zlink.samples.shoppingmall.shared.contracts.Messages;

/**
 * Self-check hook that seeds a pending idempotency mapping so the
 * pending-recovery scenario can prove forwarding to the owning instance.
 */
@ZLinkHandlerGroup("commerce")
public final class CreatePendingMappingHandler
    implements ZLinkRequestHandler<Messages.CreatePendingMappingReq, Messages.CreatePendingMappingRes> {
    private final CommerceStore store;

    public CreatePendingMappingHandler(CommerceStore store) {
        this.store = store;
    }

    @Override
    public Messages.CreatePendingMappingRes handle(
        Messages.CreatePendingMappingReq request,
        ZLinkRequestContext context) {
        store.createPendingMapping(
            request.idempotencyKey(), request.orderId(), request.ownerInstanceId());
        return new Messages.CreatePendingMappingRes(true);
    }
}

File: framework/languages/java/samples/kotlin/ShoppingMall/Server/CommerceApi/src/main/kotlin/systems/zlink/samples/kotlin/shoppingmall/server/commerceapi/handlers/CreatePendingMappingHandler.kt
package systems.zlink.samples.kotlin.shoppingmall.server.commerceapi.handlers

import systems.zlink.framework.channels.ZLinkRequestContext
import systems.zlink.framework.channels.ZLinkRequestHandler
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.samples.kotlin.shoppingmall.server.configuration.CommerceStore
import systems.zlink.samples.kotlin.shoppingmall.shared.contracts.CreatePendingMappingReq
import systems.zlink.samples.kotlin.shoppingmall.shared.contracts.CreatePendingMappingRes

/**
 * Self-check hook that seeds a pending idempotency mapping so the
 * pending-recovery scenario can prove forwarding to the owning instance.
 */
@ZLinkHandlerGroup("commerce")
class CreatePendingMappingHandler(
    private val store: CommerceStore,
) : ZLinkRequestHandler<CreatePendingMappingReq, CreatePendingMappingRes> {
    override fun handle(
        request: CreatePendingMappingReq,
        context: ZLinkRequestContext,
    ): CreatePendingMappingRes {
        store.createPendingMapping(request.idempotencyKey, request.orderId, request.ownerInstanceId)
        return CreatePendingMappingRes(true)
    }
}
